collectors/plugins.d/pluginsd_parser.c

```diff
@@ -740,7 +740,252 @@ static void pluginsd_process_thread_cleanup(void *ptr) {
     parser_destroy(parser);
 }
 
-// New plugins.d parser
+// Pluginsd_action for the replication commands
+PARSER_RC pluginsd_rep_action(void *user, REP_ARG command)
+{
+    if(!user || !((PARSER_USER_OBJECT *)user)->host || !((PARSER_USER_OBJECT *)user)->opaque) {
+        infoerr("%s: Parser user object was not set properly - user, host or opaque is NULL - Exiting Parser!", REPLICATION_MSG);
+        return PARSER_RC_ERROR;
+    }
+    RRDHOST *host = ((PARSER_USER_OBJECT *)user)->host;
+    REPLICATION_STATE *rep_state = ((PARSER_USER_OBJECT *)user)->opaque;
+
+    switch (command)
+    {
+      case REP_OFF:
+        debug(D_REPLICATION, "%s: REP OFF command received!\n", REPLICATION_MSG);
+        // Shutdown the replication thread and exit the parser
+        rep_state->shutdown = 1;
+        ((PARSER_USER_OBJECT *)user)->enabled = 0;
+        return PARSER_RC_ERROR;
+      case REP_ON:
+        debug(D_REPLICATION, "%s: REP ON command is received!\n", REPLICATION_MSG);
+        if (finish_gap_replication(host, rep_state)) {
+            ((PARSER_USER_OBJECT *)user)->enabled = 0;
+            return PARSER_RC_ERROR;
+        }
+        send_gap_for_replication(host, rep_state);
+        return PARSER_RC_OK;
+      case REP_ACK:
+        debug(D_REPLICATION, "%s: REP ACK command is received!\n", REPLICATION_MSG);
+        // REP ACK - An acknoledgement for a full REP transmission of all the charts of one GAP.
+        cleanup_after_gap_replication(host->gaps_timeline);
+        if (finish_gap_replication(host, rep_state)) {
+            ((PARSER_USER_OBJECT *)user)->enabled = 0;
+            return PARSER_RC_ERROR;
+        }
+        send_gap_for_replication(host, rep_state);
+        return PARSER_RC_OK;
+      default:
+        debug(D_REPLICATION,"%s: REP %u command is unknown!\n", REPLICATION_MSG, command);
+    }
+
+    return PARSER_RC_OK;
+}
+
+PARSER_RC pluginsd_gap_action(void *user, GAP rx_gap)
+{
+    if(!user || !((PARSER_USER_OBJECT *)user)->opaque) {
+        infoerr("%s: Parser user object was not set properly - user, host or opaque is NULL - Exiting Parser!", REPLICATION_MSG);
+        return PARSER_RC_ERROR;
+    }
+    REPLICATION_STATE *rep_state = ((PARSER_USER_OBJECT *)user)->opaque;
+    //Check if there is GAP and send GAP command, otherwise send REP OFF command
+    sender_gap_filling(rep_state, rx_gap);
+    char gap_uuid_str[UUID_STR_LEN + 1];
+    uuid_unparse(rx_gap.gap_uuid, gap_uuid_str);
+    debug(D_REPLICATION,"%s: Gap transmission completed - Sending REP ACK for GAP %s\n", REPLICATION_MSG, gap_uuid_str);
+    send_message(rep_state, "REP 3\n");
+
+    return PARSER_RC_OK;
+}
+
+PARSER_RC pluginsd_rdata_action(void *user, GAP meta_rx_rdata, int block_id, char *chart_id, char *dim_id)
+{
+#ifndef NETDATA_INTERNAL_CHECKS
+    UNUSED(block_id);
+#endif
+    if(!user || !((PARSER_USER_OBJECT *)user)->opaque) {
```

Comment:
Is there any possibility to call this function with `chart_id` and `dim_id` as NULL? If the answer is positive, we should add here `chart_id` and `dim_id`